Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -1173,6 +1185,129 @@ static String mutationObserverNotificationString()
     return scrollingCoordinator.requestStopKeyboardScrollAnimation(*this, immediate);
 }
 
+void UnifiedPDFPlugin::populateScrollSnapIdentifiers()
+{
+    ASSERT(m_scrollSnapIdentifiers.isEmpty());
+
+    // Create fake ElementIdentifiers for each page.
+    // FIXME: Ideally we would not use DOM Element identifiers here, but scroll
+    // snap code wants them (but never uses them to look the element back up).
+    // We should consider making it generic.
+    for (PDFDocumentLayout::PageIndex i = 0; i < m_documentLayout.pageCount(); ++i)
+        m_scrollSnapIdentifiers.append(ElementIdentifier::generate());
+}
+
+PDFDocumentLayout::PageIndex UnifiedPDFPlugin::pageForScrollSnapIdentifier(ElementIdentifier scrollSnapIdentifier) const
+{
+    auto index = m_scrollSnapIdentifiers.find(scrollSnapIdentifier);
+    RELEASE_ASSERT(index != notFound);
+    return index;
+}
+
+bool UnifiedPDFPlugin::shouldUseScrollSnapping() const
+{
+    return m_documentLayout.displayMode() == PDFDocumentLayout::DisplayMode::SinglePage || m_documentLayout.displayMode() == PDFDocumentLayout::DisplayMode::TwoUp;
+}
+
+void UnifiedPDFPlugin::updateSnapOffsets()
+{
+    if (m_scrollSnapIdentifiers.isEmpty())
+        populateScrollSnapIdentifiers();
+
+    if (!shouldUseScrollSnapping()) {
+        clearSnapOffsets();
+        return;
+    }
+
+    Vector<SnapOffset<LayoutUnit>> verticalSnapOffsets;
+    Vector<LayoutRect> snapAreas;
+
+    for (PDFDocumentLayout::PageIndex i = 0; i < m_documentLayout.pageCount(); ++i) {
+        auto destinationRect = m_documentLayout.boundsForPageAtIndex(i);
```

Comment:
This feels like a "page to contents" type conversion which should go through the helpers.

Reply:
yep, the helper didn't exist yet when I wrote this :)